resolver/map.go

```diff
@@ -28,25 +28,36 @@ type addressMapEntry struct {
 // Multiple accesses may not be performed concurrently.  Must be created via
 // NewAddressMap; do not construct directly.
 type AddressMap struct {
-	m map[string]addressMapEntryList
+	// The underlying map is keyed by an Address with fields that we don't care
+	// about being set to their zero values. The only fields that we care about
+	// are `Addr`, `ServerName` and `Attributes`. Since we need to be able to
+	// distinguish between addresses with same `Addr` and `ServerName`, but
+	// different `Attributes`, we cannot store the `Attributes` in the map key.
+	//
+	// The comparison operation for structs work as follows:
+	//  Struct values are comparable if all their fields are comparable. Two
+	//  struct values are equal if their corresponding non-blank fields are equal.
+	//
+	// The value type of the map contains a slice of addresses which match the key
+	// in their `Addr` and `ServerName` fields and contain the corresponding value
+	// associated with them.
```

Comment:
Great.  Now, should we change `addressMapEntry` to be an `Attributes` instead of an `Address`, since the other parts of `Address` that we can hash are in the map key?

Reply:
The problem with doing that is that as part of `Keys()` and `Get()`, we need to return the actual `resolver.Address` that the user added to our map. Without storing the full address (as given) by the user in `Set()`, there is no way to return it.